File: 53.最大子数组和.go
/*
 * @lc app=leetcode.cn id=53 lang=golang
 *
 * [53] 最大子数组和
 */

package main

// @lc code=start
func maxSubArray(nums []int) int {
	prev := 0
	maxAns := nums[0]

	for _, num := range nums {
		prev = max(num, prev+num)
		maxAns = max(prev, maxAns)
	}

	return maxAns
}

func max(a, b int) int {
	if a > b {
		return a
	} else {
		return b
	}
}

// @lc code=end

File: 53.最大子数组和.rs
/*
 * @lc app=leetcode.cn id=53 lang=rust
 *
 * [53] 最大子数组和
 */

use crate::structs::*;

// @lc code=start
impl Solution {
    pub fn max_sub_array(nums: Vec<i32>) -> i32 {
        let (mut prev, mut max_ans) = (0, nums[0]);
        for num in nums {
            prev = num.max(prev + num);
            max_ans = max_ans.max(prev);
        }
        max_ans
    }
}
// @lc code=end

